lib/os/onoff.c

```diff
@@ -40,36 +39,11 @@ static int validate_args(const struct onoff_service *srv,
 		return -EINVAL;
 	}
 
-	int rv = 0;
-	u32_t mode = cli->flags;
+	int rv = async_notify_validate(&cli->notify);
 
-	/* Reject unexpected flags. */
-	if (mode != (cli->flags & CLIENT_VALID_FLAGS_MASK)) {
-		return -EINVAL;
-	}
-
-	/* Validate configuration based on mode */
-	switch (mode & CLIENT_NOTIFY_METHOD_MASK) {
-	case ONOFF_CLIENT_NOTIFY_SPINWAIT:
-		break;
-	case ONOFF_CLIENT_NOTIFY_CALLBACK:
-		if (cli->async.callback.handler == NULL) {
-			rv = -EINVAL;
-		}
-		break;
-	case ONOFF_CLIENT_NOTIFY_SIGNAL:
-		if (cli->async.signal == NULL) {
-			rv = -EINVAL;
-		}
-		break;
-	default:
+	if ((rv == 0)
+	    && ((cli->notify.flags & ASYNC_NOTIFY_EXTENSION_MASK) != 0)) {
```

Comment:
Is this actually needed? Why not just `return async_notify_validate(&cli->notify)` here?

Reply:
This catches a situation where the client structure has not been initialized but the limited check of `async_notify_validate()` passes.  That requires only that the method be correct and consistent; it doesn't look for other bogus values.  This service doesn't currently use extension bits, so those should be zeroed if the structure was properly initialized.

A unit test fails if this check is removed: in that case the structure appears to be a valid callback configuration because the low two bits of flags identify callback and there's a non-null pointer, but in fact the pointer is garbage and a segfault occurs if its invoked.